### PR TITLE
Addressing Security Concerns Raised by OpenSSF Scorecard

### PR DESCRIPTION
**Pull request by NishkarshRaj**
## Hey, I just made a Pull Request!

Added:

- Pinned Dependency to SHA level
- Permissions defined for GitHub Actions following PoLP

Note: This PR was originally generated in author's personal fork using the standard: https://app.stepsecurity.io/ given I don't have contributor access to upstream Backstage.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by NishkarshRaj** [comment]
@freben : Unsure why the sole CI check failed. Can't make sense out of the logs - 

https://github.com/backstage/backstage/actions/runs/5784684796/job/15675859390?pr=19249

do I have to make any changes?

**Comment by adamdmharvey** [comment]
Love that you're helping contribute this one, @NishkarshRaj !

Looks like there's a conflict in one of the YAMLs, but an overall comment.  I wonder if the blast radius 💥 of this PR is perhaps a bit large.  The AppSecurity PR review feature is awesome - I've used it before. But via this PR, You're attempting to modify,

* Literally every single GitHub Action in the entire project
* Both pinning dependencies - which is great - but also setting default permissions
* And changing Dockerfile examples in a create app which gets delivered to users when they scaffold and are testing backstage for the first time.

So it's sort of like a few different MAJOR things to review, and agree on, to be able to merge in.  And a mistake to the workflows could be really impacting.

I know this may add more work, but if I could suggest, I'd break down the changes into smaller PRs.  For example, rather than taking literally EVERY SINGLE PR, maybe only firing changes to groups of them - ones that perhaps are the easiest like the NOOPS or the ones which are obvious like labeling PRs, or all of the E2E ones together, etc.  That way they are a) smaller to review and read, and b) easier to test their impact upon merging.

The repo already has reports on these OpenSSF vulnerabilities now, so as we chip away at them the security alerts will get closed out and the maintainers can track them.  And if we need we can use that AppSecurity review tool iteratively to continue to improve it.

What do you think?

**Comment by adamdmharvey** [comment]
PS: For the maintainers, I happened to recently test just to make sure, but Dependabot is smart in terms of patching that if we specify Git hash in this to help make sure there's no version drift, Dependabot is smart in that it still help us maintain by version and semver, but will maintain the git hash which is a key aspect to help protect build security and supply chain traceability.

Example: https://github.com/adamdmharvey/actions-safe-and-maintained/pull/3/files

**Comment by freben** [comment]
Hey, sorry I let this linger a bit. I added a needs-discussion tag on this to bring it up at the maintainer sync :)

**Comment by freben** [comment]
(needs a rebase now, unfortunately, as well)

**Comment by NishkarshRaj** [comment]
@adamdmharvey @freben - Thanks for the response!

I do believe the blast radius of the PR is wide but the notion was always to help improve security.

Given Backstage gets loads of update regularly, I'll keep this open until a maintainer picks it up and re-creates a PR with the App Security utility else a rebase now would become obsolete later.

Happy to discuss and contribute this in small(er) incremental checks.

Security FTW <3

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thank you for the suggestions here! 🙏 

We've discussed this quite a bit and found a couple of things. First off the constrained permissions sounds good, let's do that 👍

We're quite torn on the pinning of the actions SHA. For all of the GitHub provided actions as well as our own (`actions/*`, `github/*`, `backstage/*`) we think it's best to avoid the pinning. It seems better to quickly receive patches rather than potentially lingering on vulnerable versions. For the rest of the actions it would seem like a net benefit though, there's the issue of patch delay and we'll most likely end up merging most SHA bumps pretty robotically, but in the end we add a small manual review step which might catch something.

Regarding the `step-security/harden-runner` we're curious why they were only added to some workflows?

**Comment by NishkarshRaj** [comment]
@Rugvip : Thanks for the reply.

Basis your requested review changes, I have committed the following:

1. Added Harden Runner for all Actions file (No clue how the tool missed it for some files)
2. Removed pinned SHA dependency on all accepted sources: `github/*` , `backstage/*` , `actions/*` - do we need to extend these to established sources like Docker and Snyk as well?
3. Removed pinned SHA dependency on the base image of Dockerfile.

File: .changeset/rare-files-crash.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/create-app': patch
+---
+
+Addressing Security concerns raised by OpenSSF Scorecard checks - pinned dependency and scoped permission model added for GitHub Actions workflow
```
[the remark this answers is not in the document]
Comment (NishkarshRaj) [reply .changeset/rare-files-crash.md]: Thanks! Nice shout

**Comment by NishkarshRaj** [comment]
I removed the extraneous changeset and rebased with master last night - don't know why the build is failing as my changes have not affected any code base per se.

I'll do another rebase with the latest master now but think something @Rugvip @freben you can help me understand better?

**Comment by NishkarshRaj** [comment]
@Rugvip @freben Unsure what is breaking the jobs.

There are some markdown references seemingly broken but given I took a merge from the master branch they should have been resolved?

**Comment by NishkarshRaj** [comment]
Now something strange with the Discord noti. Any nudges @freben @adamdmharvey @Rugvip ?

**Comment by adamdmharvey** [comment]
I reran the microsite action this morning (US Eastern), in case it was intermittent. Seems it ran clean so that's good ! Discord notification is usually used only for failed runs ? I'll take a look in a bit to see if we can rerun that or clear it.

**Comment by NishkarshRaj** [comment]
@adamdmharvey @Rugvip - any nudges to clean this up and get it shipped? :shipit:

**Comment by adamdmharvey** [comment]
Thanks for the nudge; re-ran the build job. Let's see!

**Comment by NishkarshRaj** [comment]
@Rugvip @adamdmharvey - any chance this failing workflow is related to #19984?

**Comment by NishkarshRaj** [comment]
Welp! Some stunts to somehow fix the broken pipeline hell caused me to lose the history - reopened another PR to address these changes here: #20090 20090

cc/ @Rugvip @adamdmharvey @freben

**Pull request closed without merging**